discovery/discovery.go

```diff
@@ -10,23 +13,80 @@ import (
 	"github.com/golang/glog"
 )
 
-type offchainOrchestrator struct {
-	uri   *url.URL
+const GetOrchestratorsTimeoutLoop = 3 * time.Second
+
+type offchainOrchestrators struct {
+	uri   []*url.URL
 	bcast server.Broadcaster
 }
 
-func NewOffchainOrchestrator(node *core.LivepeerNode, address string) *offchainOrchestrator {
+func NewOffchainOrchestrator(node *core.LivepeerNode, address string) *offchainOrchestrators {
 	addr := "https://" + address
 	uri, err := url.ParseRequestURI(addr)
 	if err != nil {
 		glog.Error("Could not parse orchestrator URI: ", err)
 		return nil
 	}
 	bcast := core.NewBroadcaster(node)
-	return &offchainOrchestrator{bcast: bcast, uri: uri}
+	return &offchainOrchestrators{bcast: bcast, uri: []*url.URL{uri}}
+}
+
+func NewOffchainOrchestratorFromOnchainList(node *core.LivepeerNode) *offchainOrchestrators {
+	orchestrators, err := node.Eth.RegisteredTranscoders()
+	if err != nil {
+		glog.Error(err)
+		return nil
+	}
+
+	var uris []*url.URL
+
+	for _, orch := range orchestrators {
+		serviceURI := orch.ServiceURI
+		if !strings.HasPrefix(serviceURI, "http") {
+			serviceURI = "https://" + serviceURI
+		}
+		uri, err := url.ParseRequestURI(serviceURI)
+		if err != nil {
+			glog.Error("Could not parse orchestrator URI: ", err)
+			return nil
+		}
+		uris = append(uris, uri)
+	}
+
+	bcast := core.NewBroadcaster(node)
+	return &offchainOrchestrators{bcast: bcast, uri: uris}
 }
 
-func (o *offchainOrchestrator) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
-	tinfo, err := server.GetOrchestratorInfo(o.bcast, o.uri)
-	return []*net.OrchestratorInfo{tinfo}, err
+func (o *offchainOrchestrators) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), GetOrchestratorsTimeoutLoop)
+
+	var orchInfo []*net.OrchestratorInfo
+	orchChan := make(chan struct{})
+
+	getOrchInfo := func(uri *url.URL, bcast server.Broadcaster, i int) {
+		tinfo, err := server.GetOrchestratorInfo(bcast, uri)
+		if err != nil {
+			glog.Errorf("Error getting OrchestratorInfo for uri: %v", uri)
+			return
+		}
+		orchInfo = append(orchInfo, tinfo)
+		if numOrchestrators <= i {
```

Comment:
This probably won't work as expected; it'd signal the chan for the `i`th request, regardless of whether the other requests have completed. I think something like this will work for `GetOrchestrators`:

```golang
numResp := 0
numSuccessResp := 0
respLock := sync.Mutex{}
orchInfos := []*net.OrchestratorInfo{}
getOrchInfo := func(uri *url.URL) {
   // should we also modify this to take a cancelable context?
  info, err = server.GetOrchestratorInfo(o.bcast, uri)
  respLock.Lock()
  defer respLock.Unlock()
  numResp++
  if err != nil {
    orchInfos = append(orchInfos, info)
    numSuccessResp++
  }
  // Signal the chan if we've received the expected number of successful responses
  // or if we've exhausted the list of URIs
  // If we ever look at other metrics for selecting O, then we'll need to gather
  // *all* responses, rather than stop at numOrchestrators. But this is OK for now.
  if numSuccessResp >= numOrchestrators || numResp >= len(o.uris) {
    orchChan <- struct{}{}
  }
}
for _, uri := range o.uris {
  go getOrchInfo(uri)
}
select {
  case <- ctx.Done():
    ...
  case <- orchChan:
    ...
}
```

Reply:
Just realized there's a typo here -- the `err  != nil` check should be `err == nil` !